File: WiFiMonitor.cs
// WiFiMonitor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace WiFiMonitor
{
    class Result
    {
        public double? Ping { get; set; }
        public double? Download { get; set; }
        public double? Upload { get; set; }
    }

    class Monitor
    {
        private string pingHost;
        private int downloadSize;
        private int uploadSize;
        private int count;
        private bool color;
        private List<Result> results = new List<Result>();
        private HttpClient client = new HttpClient();

        public Monitor(string pingHost, int downloadSize, int uploadSize, int count, bool color)
        {
            this.pingHost = pingHost;
            this.downloadSize = downloadSize;
            this.uploadSize = uploadSize;
            this.count = count;
            this.color = color || Console.IsOutputRedirected == false;
        }

        private double? MeasurePing(string host)
        {
            try
            {
                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "ping",
                        Arguments = $"-c 4 {host}",
                        RedirectStandardOutput = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };
                process.Start();
                string output = process.StandardOutput.ReadToEnd();
                process.WaitForExit();
                foreach (var line in output.Split('\n'))
                {
                    if (line.Contains("avg") || line.Contains("rtt min/avg/max/mdev"))
                    {
                        var parts = line.Split('=');
                        if (parts.Length > 1)
                        {
                            var nums = parts[1].Split('/');
                            if (nums.Length > 1 && double.TryParse(nums[1], out double avg))
                                return avg;
                        }
                    }
                }
            }
            catch { }
            return null;
        }

        private double? MeasureDownload()
        {
            string url = "http://ipv4.download.thinkbroadband.com/5MB.zip";
            try
            {
                var start = DateTime.Now;
                var response = client.GetAsync(url).Result;
                long total = 0;
                var stream = response.Content.ReadAsStreamAsync().Result;
                byte[] buffer = new byte[8192];
                int read;
                while ((read = stream.Read(buffer, 0, buffer.Length)) > 0)
                    total += read;
                double elapsed = (DateTime.Now - start).TotalSeconds;
                return (total * 8) / (elapsed * 1_000_000);
            }
            catch { return null; }
        }

        private double? MeasureUpload()
        {
            string url = "https://httpbin.org/post";
            int size = uploadSize * 1024 * 1024;
            byte[] data = new byte[size];
            for (int i = 0; i < data.Length; i++) data[i] = 48; // '0'
            try
            {
                var start = DateTime.Now;
                var content = new ByteArrayContent(data);
                content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/octet-stream");
                var response = client.PostAsync(url, content).Result;
                double elapsed = (DateTime.Now - start).TotalSeconds;
                return (size * 8) / (elapsed * 1_000_000);
            }
            catch { return null; }
        }

        private Result RunTest()
        {
            var res = new Result();
            if (!string.IsNullOrEmpty(pingHost)) res.Ping = MeasurePing(pingHost);
            if (downloadSize > 0) res.Download = MeasureDownload();
            if (uploadSize > 0) res.Upload = MeasureUpload();
            return res;
        }

        private void PrintResult(Result res)
        {
            if (color)
            {
                if (res.Ping.HasValue) Console.ForegroundColor = ConsoleColor.Green;
                else Console.ResetColor();
                if (res.Ping.HasValue) Console.WriteLine($"🌐 Ping: {res.Ping.Value:F2} ms");
                if (res.Download.HasValue) { Console.ForegroundColor = ConsoleColor.Blue; Console.WriteLine($"📥 Download: {res.Download.Value:F2} Mbps"); }
                if (res.Upload.HasValue) { Console.ForegroundColor = ConsoleColor.Cyan; Console.WriteLine($"📤 Upload: {res.Upload.Value:F2} Mbps"); }
                Console.ResetColor();
            }
            else
            {
                if (res.Ping.HasValue) Console.WriteLine($"Ping: {res.Ping.Value:F2} ms");
                if (res.Download.HasValue) Console.WriteLine($"Download: {res.Download.Value:F2} Mbps");
                if (res.Upload.HasValue) Console.WriteLine($"Upload: {res.Upload.Value:F2} Mbps");
            }
            Console.WriteLine();
        }

        public async Task RunAsync()
        {
            for (int i = 0; i < count; i++)
            {
                Console.WriteLine($"Тест {i+1}/{count}");
                var res = RunTest();
                results.Add(res);
                PrintResult(res);
                if (i < count - 1) await Task.Delay(1000);
            }
        }

        public void Save(string filename)
        {
            string ext = Path.GetExtension(filename).ToLower().TrimStart('.');
            string content;
            if (ext == "json")
                content = JsonConvert.SerializeObject(results, Formatting.Indented);
            else if (ext == "csv")
            {
                var sb = new StringBuilder("ping,download,upload\n");
                foreach (var r in results)
                    sb.AppendLine($"{r.Ping},{r.Download},{r.Upload}");
                content = sb.ToString();
            }
            else
                content = string.Join("\n", results);
            File.WriteAllText(filename, content);
            Console.WriteLine($"Результаты сохранены в {filename}");
        }

        public static async Task Main(string[] args)
        {
            string pingHost = null;
            int downloadSize = 0, uploadSize = 0, count = 1;
            string output = null;
            bool color = false;

            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--ping": pingHost = args[++i]; break;
                    case "--download": downloadSize = int.Parse(args[++i]); break;
                    case "--upload": uploadSize = int.Parse(args[++i]); break;
                    case "--count": count = int.Parse(args[++i]); break;
                    case "--output": output = args[++i]; break;
                    case "--color": color = true; break;
                }
            }
            if (pingHost == null && downloadSize == 0 && uploadSize == 0)
            {
                pingHost = "8.8.8.8";
                downloadSize = 5;
                uploadSize = 2;
            }
            var monitor = new Monitor(pingHost, downloadSize, uploadSize, count, color);
            await monitor.RunAsync();
            if (output != null) monitor.Save(output);
        }
    }
}

File: wifi_monitor.go
// wifi_monitor.go
package main

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

type Result struct {
	Ping     *float64 `json:"ping,omitempty"`
	Download *float64 `json:"download,omitempty"`
	Upload   *float64 `json:"upload,omitempty"`
}

type Monitor struct {
	pingHost     string
	downloadSize int
	uploadSize   int
	count        int
	color        bool
	results      []Result
}

func NewMonitor(pingHost string, downloadSize, uploadSize, count int, color bool) *Monitor {
	return &Monitor{
		pingHost:     pingHost,
		downloadSize: downloadSize,
		uploadSize:   uploadSize,
		count:        count,
		color:        color,
	}
}

func (m *Monitor) measurePing(host string) *float64 {
	cmd := exec.Command("ping", "-c", "4", host)
	out, err := cmd.Output()
	if err != nil {
		return nil
	}
	lines := strings.Split(string(out), "\n")
	for _, line := range lines {
		if strings.Contains(line, "avg") || strings.Contains(line, "rtt min/avg/max/mdev") {
			parts := strings.Split(line, "=")
			if len(parts) > 1 {
				fields := strings.Split(parts[1], "/")
				if len(fields) > 1 {
					avg, err := strconv.ParseFloat(fields[1], 64)
					if err == nil {
						return &avg
					}
				}
			}
		}
	}
	return nil
}

func (m *Monitor) measureDownload() *float64 {
	url := "http://ipv4.download.thinkbroadband.com/5MB.zip"
	start := time.Now()
	resp, err := http.Get(url)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()
	var total int64
	buf := make([]byte, 8192)
	for {
		n, err := resp.Body.Read(buf)
		if err != nil && err != io.EOF {
			return nil
		}
		total += int64(n)
		if err == io.EOF {
			break
		}
	}
	elapsed := time.Since(start).Seconds()
	speed := float64(total*8) / (elapsed * 1_000_000)
	return &speed
}

func (m *Monitor) measureUpload() *float64 {
	url := "https://httpbin.org/post"
	size := m.uploadSize * 1024 * 1024
	data := bytes.Repeat([]byte("0"), size)
	start := time.Now()
	_, err := http.Post(url, "application/octet-stream", bytes.NewReader(data))
	if err != nil {
		return nil
	}
	elapsed := time.Since(start).Seconds()
	speed := float64(size*8) / (elapsed * 1_000_000)
	return &speed
}

func (m *Monitor) runTest() Result {
	res := Result{}
	if m.pingHost != "" {
		res.Ping = m.measurePing(m.pingHost)
	}
	if m.downloadSize > 0 {
		res.Download = m.measureDownload()
	}
	if m.uploadSize > 0 {
		res.Upload = m.measureUpload()
	}
	return res
}

func (m *Monitor) printResult(res Result) {
	if m.color {
		if res.Ping != nil {
			fmt.Printf("\033[32m🌐 Ping: %.2f ms\033[0m\n", *res.Ping)
		}
		if res.Download != nil {
			fmt.Printf("\033[34m📥 Download: %.2f Mbps\033[0m\n", *res.Download)
		}
		if res.Upload != nil {
			fmt.Printf("\033[36m📤 Upload: %.2f Mbps\033[0m\n", *res.Upload)
		}
	} else {
		if res.Ping != nil {
			fmt.Printf("Ping: %.2f ms\n", *res.Ping)
		}
		if res.Download != nil {
			fmt.Printf("Download: %.2f Mbps\n", *res.Download)
		}
		if res.Upload != nil {
			fmt.Printf("Upload: %.2f Mbps\n", *res.Upload)
		}
	}
	fmt.Println()
}

func (m *Monitor) Run() {
	for i := 0; i < m.count; i++ {
		fmt.Printf("Тест %d/%d\n", i+1, m.count)
		res := m.runTest()
		m.results = append(m.results, res)
		m.printResult(res)
		if i < m.count-1 {
			time.Sleep(1 * time.Second)
		}
	}
}

func (m *Monitor) Save(filename string) error {
	ext := strings.ToLower(filename[strings.LastIndex(filename, ".")+1:])
	var data []byte
	switch ext {
	case "json":
		d, err := json.MarshalIndent(m.results, "", "  ")
		if err != nil {
			return err
		}
		data = d
	case "csv":
		var buf bytes.Buffer
		w := csv.NewWriter(&buf)
		w.Write([]string{"ping", "download", "upload"})
		for _, r := range m.results {
			row := []string{
				fmt.Sprintf("%v", r.Ping),
				fmt.Sprintf("%v", r.Download),
				fmt.Sprintf("%v", r.Upload),
			}
			w.Write(row)
		}
		w.Flush()
		data = buf.Bytes()
	default:
		// text
		var s string
		for _, r := range m.results {
			s += fmt.Sprintf("%v\n", r)
		}
		data = []byte(s)
	}
	return os.WriteFile(filename, data, 0644)
}

func main() {
	var (
		pingHost     string
		downloadSize int
		uploadSize   int
		count        int
		output       string
		color        bool
	)
	flag.StringVar(&pingHost, "ping", "", "Хост для пинга")
	flag.IntVar(&downloadSize, "download", 0, "Размер для загрузки в МБ")
	flag.IntVar(&uploadSize, "upload", 0, "Размер для отдачи в МБ")
	flag.IntVar(&count, "count", 1, "Количество тестов")
	flag.StringVar(&output, "output", "", "Файл для сохранения")
	flag.BoolVar(&color, "color", false, "Принудительно включить цвет")
	flag.Parse()

	if pingHost == "" && downloadSize == 0 && uploadSize == 0 {
		pingHost = "8.8.8.8"
		downloadSize = 5
		uploadSize = 2
	}
	monitor := NewMonitor(pingHost, downloadSize, uploadSize, count, color || isTerminal())
	monitor.Run()
	if output != "" {
		if err := monitor.Save(output); err != nil {
			fmt.Fprintf(os.Stderr, "Ошибка сохранения: %v\n", err)
		} else {
			fmt.Printf("Результаты сохранены в %s\n", output)
		}
	}
}

func isTerminal() bool {
	fileInfo, _ := os.Stdout.Stat()
	return (fileInfo.Mode() & os.ModeCharDevice) != 0
}
